s2i/builder.go

```diff
@@ -0,0 +1,115 @@
+package s2i
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	dockerClient "github.com/docker/docker/client"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/api/validation"
+	"github.com/openshift/source-to-image/pkg/build"
+	"github.com/openshift/source-to-image/pkg/build/strategies"
+	"github.com/openshift/source-to-image/pkg/scm/git"
+
+	fn "knative.dev/kn-plugin-func"
+	docker "knative.dev/kn-plugin-func/docker"
+)
+
+var (
+	// ErrRuntimeRequired indicates the required value of Function Runtime was not provided
+	ErrRuntimeRequired = errors.New("runtime is required to build")
+
+	// ErrRuntimeNotSupported indicates the given runtime is not (yet) supported
+	// by this builder.
+	ErrRuntimeNotSupported = errors.New("runtime not supported")
+)
+
+// DefaultBuilderImages for s2i builders indexed by Runtime Language
+var DefaultBuilderImages = map[string]string{
+	"node": "registry.access.redhat.com/ubi8/nodejs-16", // TODO: finalize choice and include version
+}
+
+// Builder of Functions using the s2i subsystem.
+type Builder struct {
+	verbose bool
+}
+
+// NewBuilder creates a new instance of a Builder with static defaults.
+func NewBuilder(verbose bool) *Builder {
+	return &Builder{verbose: verbose}
+}
+
+func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
+	// Ensure the Function has a builder specified
+	if f.Builder == "" {
+		f.Builder, err = defaultBuilderImage(f)
+		if err != nil {
+			return
+		}
+	}
+
+	client, endpoint, err := docker.NewClient(dockerClient.DefaultDockerHost)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	// TODO: this defaulting of endpoint should perhaps not be necessary?  The
+	// default is explicitly passed to NewClient, thus the returned endpoint
+	// should at least be that.
+	if endpoint == "" {
+		endpoint = dockerClient.DefaultDockerHost
+	}
+
+	// Build Config
+	cfg := &api.Config{}
+	cfg.Quiet = !b.verbose
+	cfg.Tag = f.Image
+	cfg.Source = &git.URL{URL: url.URL{Path: f.Root}, Type: git.URLTypeLocal}
+	cfg.BuilderImage = f.Builder
+	cfg.BuilderPullPolicy = api.DefaultBuilderPullPolicy
+	cfg.PreviousImagePullPolicy = api.DefaultPreviousImagePullPolicy
+	cfg.RuntimeImagePullPolicy = api.DefaultRuntimeImagePullPolicy
+	cfg.DockerConfig = &api.DockerConfig{
+		Endpoint: endpoint,
+	}
+	if errs := validation.ValidateConfig(cfg); len(errs) > 0 {
+		for _, e := range errs {
+			fmt.Fprintf(os.Stderr, "ERROR: %s\n", e)
+		}
+		return errors.New("Unable to build via the s2i builder.")
+	}
+
+	builder, _, err := strategies.Strategy(client, cfg, build.Overrides{})
```

Comment:
Am I correct to understand this is the "docker" build strategy? One of the other build strategies that s2i uses is "binary" which we may consider leveraging for push-to-cluster-my-function-and-build-it-there, since that's exactly what it does.

Reply:
Yes, it is my understanding this is the Docker method.  I am not familiar with the binary strategy, but I will look into it.